src/libAtomVM/otp_socket.c

```diff
@@ -0,0 +1,1356 @@
+        }
+    }
+}
+
+//
+// peername
+//
+
+static void do_peername(Context *ctx, term msg)
+{
+    TRACE("do_peername\n");
+    GlobalContext *global = ctx->global;
+    struct SocketData *socket_data = (struct SocketData *) ctx->platform_data;
+
+    term pid = term_get_tuple_element(msg, 0);
+    term ref = term_get_tuple_element(msg, 1);
+    uint64_t ref_ticks = term_to_ref_ticks(ref);
+
+    struct sockaddr_in addr;
+    socklen_t addrlen = sizeof(addr);
+    int res = getpeername(socket_data->fd, (struct sockaddr *) &addr, &addrlen);
+
+    if (res != 0) {
+        AVM_LOGE(TAG, "Unable to getpeername: fd=%i res=%i.", socket_data->fd, res);
+        sync_send_error_tuple(ctx, pid, ref_ticks, map_errno(errno, global));
+    } else {
+        // {Ref :: reference(), {ok, #{addr => {a,b,c,d}, port => integer()}}}
+        if (UNLIKELY(memory_ensure_free(ctx, TUPLE_SIZE(2) + REF_SIZE + TUPLE_SIZE(2) + term_map_size_in_terms(2) + TUPLE_SIZE(4)) != MEMORY_GC_OK)) {
+            AVM_LOGW(TAG, "Failed to allocate memory: %s:%i.\n", __FILE__, __LINE__);
+            sync_send_message(pid, OUT_OF_MEMORY_ATOM, global);
+        } else {
+            term address = socket_tuple_from_addr(ctx, ntohl(addr.sin_addr.s_addr));
+            term port_number = term_from_int(ntohs(addr.sin_port));
+
+            term map = term_alloc_map(2, &ctx->heap);
+            term_set_map_assoc(map, 0, ADDR_ATOM, address);
+            term_set_map_assoc(map, 1, PORT_ATOM, port_number);
+            term return_value = port_create_tuple2(ctx, OK_ATOM, map);
+
+            sync_send_reply(ctx, pid, ref_ticks, return_value);
+        }
+    }
+}
+
+//
+// setopt
+//
+
+static void do_setopt(Context *ctx, term msg, term cmd)
+{
+    TRACE("do_setopt\n");
+    GlobalContext *global = ctx->global;
+    struct SocketData *socket_data = (struct SocketData *) ctx->platform_data;
+
+    term pid = term_get_tuple_element(msg, 0);
+    term ref = term_get_tuple_element(msg, 1);
+    uint64_t ref_ticks = term_to_ref_ticks(ref);
+    term level_tuple = term_get_tuple_element(cmd, 1);
+    term value = term_get_tuple_element(cmd, 2);
+
+    term opt = term_get_tuple_element(level_tuple, 1);
+    if (globalcontext_is_term_equal_to_atom_string(global, opt, reuseaddr_atom)) {
+        int option_value = (value == TRUE_ATOM) ? 1 : 0;
+        int res = setsockopt(socket_data->fd, SOL_SOCKET, SO_REUSEADDR, &option_value, sizeof(int));
+        if (res != 0) {
+            sync_send_error_tuple(ctx, pid, ref_ticks, map_errno(errno, global));
+        } else {
+            sync_send_reply(ctx, pid, ref_ticks, OK_ATOM);
+        }
+    } else if (globalcontext_is_term_equal_to_atom_string(global, opt, linger_atom)) {
+        term onoff = interop_kv_get_value(value, onoff_atom, ctx->global);
+        term linger = interop_kv_get_value(value, linger_atom, ctx->global);
+        struct linger sl;
+        sl.l_onoff = (onoff == TRUE_ATOM) ? 1 : 0;
```

Comment:
onoff == TRUE_ATOM would be enough.

And also we may want to make it inlined `term_to_bool`.